NPE in convertToTimestampTimeZone on timestamp comparison

Version 2.3.232

Getting an NPE on select including some timestamp comparison:
```java
Caused by: java.lang.NullPointerException: Cannot invoke "org.h2.engine.CastDataProvider.currentTimeZone()" because "<parameter2>" is null
	at org.h2.value.Value.convertToTimestampTimeZone(Value.java:2067) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.Value.convertTo(Value.java:1181) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.Value.convertTo(Value.java:1020) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.Value.compareToNotNullable(Value.java:2672) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.Value.compareTo(Value.java:2646) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.ValueRow.compareTypeSafe(ValueRow.java:128) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.Value.compareToNotNullable(Value.java:2675) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.Value.compareTo(Value.java:2646) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.CompareMode.compare(CompareMode.java:280) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.CompareMode.compare(CompareMode.java:20) ~[h2-2.3.232.jar:2.3.232]
	at java.base/java.util.TreeMap.getEntryUsingComparator(TreeMap.java:374) ~[na:na]
	at java.base/java.util.TreeMap.getEntry(TreeMap.java:344) ~[na:na]
	at java.base/java.util.TreeMap.get(TreeMap.java:279) ~[na:na]
	at org.h2.result.LocalResult.containsDistinct(LocalResult.java:249) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.expression.condition.ConditionInQuery.getValue(ConditionInQuery.java:95) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.expression.condition.ConditionInQuery.getValue(ConditionInQuery.java:55) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.expression.Expression.getBooleanValue(Expression.java:343) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.expression.SearchedCase.getValue(SearchedCase.java:30) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.command.query.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1897) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.result.LazyResult.hasNext(LazyResult.java:78) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.result.FetchedResult.next(FetchedResult.java:34) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.command.query.Select.queryFlat(Select.java:767) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.command.query.Select.queryWithoutCache(Select.java:891) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.command.query.Query.queryWithoutCacheLazyCheck(Query.java:211) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.command.query.Query.query(Query.java:530) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.command.query.Query.query(Query.java:498) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.command.CommandContainer.query(CommandContainer.java:222) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.command.Command.executeQuery(Command.java:196) ~[h2-2.3.232.jar:2.3.232]
	... 27 common frames omitted
```

The subject table:
```sql
CREATE TABLE TEST_ENTITY (
    ID VARCHAR(256) PRIMARY KEY,
    P15 TIMESTAMP(3) WITH TIME ZONE,
);
```
An NPE occures during the following select:
```sql
SELECT
  CASE
    WHEN t0.ts = ANY (
      SELECT ?::TIMESTAMP
    ) THEN TRUE
    ELSE FALSE
  END
FROM
  TEST_ENTITY t0
```

Java example to reproduce:
```java
try (Connection connection = dataSource.getConnection()){
    // Table
    PreparedStatement ddl = connection.prepareStatement("CREATE TABLE TEST_ENTITY (id VARCHAR(256) PRIMARY KEY, ts TIMESTAMP(3) with time zone)");
    ddl.execute();

    // Insert
    PreparedStatement insert = connection.prepareStatement("INSERT INTO TEST_ENTITY (id, ts) VALUES (?, ?)");
    insert.setString(1, UUID.randomUUID().toString());
    insert.setObject(2, OffsetDateTime.now());
    insert.executeUpdate();

    // NPE Select
    PreparedStatement select = connection.prepareStatement("""
        SELECT
          CASE
            WHEN t0.ts = ANY (
              SELECT ?::TIMESTAMP
            ) THEN TRUE
            ELSE FALSE
          END
        FROM
          TEST_ENTITY t0
    """);
    select.setObject(1, LocalDateTime.now());
    select.executeQuery();
} catch (SQLException e) {
    throw new RuntimeException(e);
}
```

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,12 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #4152: NPE in convertToTimestampTimeZone on timestamp comparison
+</li>
+<li>PR #4150: fix:Fix some errors in the javadoc
+</li>
+<li>PR #4149: fix:Possible overflow issues with CacheLongKeyLIRS.getMisses()
+</li>
 <li>Issue #4144: [2.3.232] Regression in ORDER BY ... DESC with WHERE ... IN and OR clauses
 </li>
 <li>Issue #4139: Oracle compatibility mode: unexpected Column alias is not specified exception in CTE

--- a/h2/src/main/org/h2/command/query/SelectGroups.java
+++ b/h2/src/main/org/h2/command/query/SelectGroups.java
@@ -72,7 +72,7 @@ public abstract class SelectGroups {
         @Override
         public void reset() {
             super.reset();
-            groupByData = new TreeMap<>(session.getDatabase().getCompareMode());
+            groupByData = new TreeMap<>(session);
             currentGroupsKey = null;
             cursor = null;
         }
@@ -346,7 +346,7 @@ public abstract class SelectGroups {
         } else {
             TreeMap<Value, PartitionData> map = windowPartitionData.get(expr);
             if (map == null) {
-                map = new TreeMap<>(session.getDatabase().getCompareMode());
+                map = new TreeMap<>(session);
                 windowPartitionData.put(expr, map);
             }
             map.put(partitionKey, obj);

--- a/h2/src/main/org/h2/engine/SessionLocal.java
+++ b/h2/src/main/org/h2/engine/SessionLocal.java
@@ -9,6 +9,7 @@ import java.time.Instant;
 import java.util.ArrayList;
 import java.util.BitSet;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -73,7 +74,7 @@ import org.h2.value.lob.LobDataInMemory;
  * mode, this object resides on the server side and communicates with a
  * SessionRemote object on the client side.
  */
-public final class SessionLocal extends Session implements TransactionStore.RollbackListener {
+public final class SessionLocal extends Session implements TransactionStore.RollbackListener, Comparator<Value> {
 
     public enum State { INIT, RUNNING, BLOCKED, SLEEP, THROTTLED, SUSPENDED, CLOSED }
 
@@ -1979,6 +1980,7 @@ public final class SessionLocal extends Session implements TransactionStore.Roll
      * @return 0 if both values are equal, -1 if the first value is smaller, and
      *         1 otherwise
      */
+    @Override
     public int compare(Value a, Value b) {
         return a.compareTo(b, this, getCompareMode());
     }

--- a/h2/src/main/org/h2/expression/aggregate/Aggregate.java
+++ b/h2/src/main/org/h2/expression/aggregate/Aggregate.java
@@ -248,7 +248,7 @@ public class Aggregate extends AbstractAggregate implements ExpressionWithFlags 
         Arrays.sort(array,
                 sortOrder != null
                         ? (v1, v2) -> sortOrder.compare(((ValueRow) v1).getList(), ((ValueRow) v2).getList())
-                        : select.getSession().getDatabase().getCompareMode());
+                        : select.getSession());
     }
 
     @Override

--- a/h2/src/main/org/h2/expression/aggregate/AggregateDataCollecting.java
+++ b/h2/src/main/org/h2/expression/aggregate/AggregateDataCollecting.java
@@ -90,7 +90,7 @@ final class AggregateDataCollecting extends AggregateData implements Iterable<Va
         Collection<Value> c = values;
         if (c == null) {
             if (distinct) {
-                Comparator<Value> comparator = session.getDatabase().getCompareMode();
+                Comparator<Value> comparator = session;
                 if (orderedWithOrder) {
                     comparator = Comparator.comparing(t -> ((ValueRow) t).getList()[0], comparator);
                 }

--- a/h2/src/main/org/h2/expression/aggregate/AggregateDataDistinctWithCounts.java
+++ b/h2/src/main/org/h2/expression/aggregate/AggregateDataDistinctWithCounts.java
@@ -42,7 +42,7 @@ final class AggregateDataDistinctWithCounts extends AggregateData {
             return;
         }
         if (values == null) {
-            values = new TreeMap<>(session.getDatabase().getCompareMode());
+            values = new TreeMap<>(session);
         }
         LongDataCounter a = values.get(v);
         if (a == null) {

--- a/h2/src/main/org/h2/expression/aggregate/Percentile.java
+++ b/h2/src/main/org/h2/expression/aggregate/Percentile.java
@@ -107,7 +107,7 @@ final class Percentile {
     static Value getValue(SessionLocal session, Value[] array, int dataType, ArrayList<QueryOrderBy> orderByList,
             BigDecimal percentile, boolean interpolate) {
         final CompareMode compareMode = session.getDatabase().getCompareMode();
-        Arrays.sort(array, compareMode);
+        Arrays.sort(array, session);
         int count = array.length;
         boolean reverseIndex = orderByList != null && (orderByList.get(0).sortType & SortOrder.DESCENDING) != 0;
         BigDecimal fpRow = BigDecimal.valueOf(count - 1).multiply(percentile);

--- a/h2/src/main/org/h2/expression/condition/ConditionInConstantSet.java
+++ b/h2/src/main/org/h2/expression/condition/ConditionInConstantSet.java
@@ -53,7 +53,7 @@ public final class ConditionInConstantSet extends ConditionIn {
     ConditionInConstantSet(SessionLocal session, Expression left, boolean not, boolean whenOperand,
             ArrayList<Expression> valueList) {
         super(left, not, whenOperand, valueList);
-        this.valueSet = new TreeSet<>(session.getDatabase().getCompareMode());
+        this.valueSet = new TreeSet<>(session);
         TypeInfo type = left.getType();
         for (Expression expression : valueList) {
             type = TypeInfo.getHigherType(type, expression.getType());

--- a/h2/src/main/org/h2/index/IndexCondition.java
+++ b/h2/src/main/org/h2/index/IndexCondition.java
@@ -191,7 +191,7 @@ public class IndexCondition {
             SortOrder sortOrder = SortOrder.ofSortTypes(session, sortTypes);
             comparator = (o1, o2) -> sortOrder.compare(((ValueRow) o1).getList(), ((ValueRow) o2).getList());
         } else {
-            comparator = session.getDatabase().getCompareMode();
+            comparator = session;
             if ((sortTypes[0] & SortOrder.DESCENDING) != 0) {
                 comparator = comparator.reversed();
             }

--- a/h2/src/main/org/h2/result/LocalResult.java
+++ b/h2/src/main/org/h2/result/LocalResult.java
@@ -206,7 +206,7 @@ public class LocalResult implements ResultInterface, ResultTarget {
     public void setDistinct() {
         assert distinctIndexes == null;
         distinct = true;
-        distinctRows = new TreeMap<>(session.getDatabase().getCompareMode());
+        distinctRows = new TreeMap<>(session);
     }
 
     /**
@@ -217,7 +217,7 @@ public class LocalResult implements ResultInterface, ResultTarget {
     public void setDistinct(int[] distinctIndexes) {
         assert !distinct;
         this.distinctIndexes = distinctIndexes;
-        distinctRows = new TreeMap<>(session.getDatabase().getCompareMode());
+        distinctRows = new TreeMap<>(session);
     }
 
     /**
@@ -227,7 +227,7 @@ public class LocalResult implements ResultInterface, ResultTarget {
      */
     public void setInPredicateValueListResult(int[] inPredicateSortTypes) {
         distinct = true;
-        distinctRows = new TreeMap<>(session.getDatabase().getCompareMode());
+        distinctRows = new TreeMap<>(session);
         if (inPredicateSortTypes.length != 0) {
             sort = SortOrder.ofSortTypes(session, inPredicateSortTypes);
         }
@@ -252,7 +252,7 @@ public class LocalResult implements ResultInterface, ResultTarget {
             return external.contains(values);
         }
         if (distinctRows == null) {
-            distinctRows = new TreeMap<>(session.getDatabase().getCompareMode());
+            distinctRows = new TreeMap<>(session);
             for (Value[] row : rows) {
                 ValueRow array = getDistinctRow(row);
                 distinctRows.put(array, array.getList());

--- a/h2/src/main/org/h2/value/CompareMode.java
+++ b/h2/src/main/org/h2/value/CompareMode.java
@@ -7,7 +7,6 @@ package org.h2.value;
 
 import java.nio.charset.Charset;
 import java.text.Collator;
-import java.util.Comparator;
 import java.util.Locale;
 import java.util.Objects;
 
@@ -17,7 +16,7 @@ import org.h2.util.StringUtils;
  * Instances of this class can compare strings. Case sensitive and case
  * insensitive comparison is supported, and comparison using a collator.
  */
-public class CompareMode implements Comparator<Value> {
+public class CompareMode {
 
     /**
      * This constant means there is no collator set, and the default string
@@ -273,11 +272,6 @@ public class CompareMode implements Comparator<Value> {
         result = 31 * result + getName().hashCode();
         result = 31 * result + strength;
         return result;
-    }
-
-    @Override
-    public int compare(Value o1, Value o2) {
-        return o1.compareTo(o2, null, this);
     }
 
 }

--- a/h2/src/test/org/h2/test/jdbc/TestPreparedStatement.java
+++ b/h2/src/test/org/h2/test/jdbc/TestPreparedStatement.java
@@ -120,6 +120,7 @@ public class TestPreparedStatement extends TestDb {
         testParameterInSubquery(conn);
         testAfterRollback(conn);
         testUnnestWithArrayParameter(conn);
+        testDateTimeWithParameter(conn);
         conn.close();
         testPreparedStatementWithLiteralsNone();
         testPreparedStatementWithIndexedParameterAndLiteralsNone();
@@ -1817,6 +1818,18 @@ public class TestPreparedStatement extends TestDb {
         ResultSet rs = prep.executeQuery();
         assertTrue(rs.next());
         assertEquals(new Integer[] { 1, 2 }, rs.getObject(1, Integer[].class));
+    }
+
+    private void testDateTimeWithParameter(Connection conn) throws SQLException {
+        Statement stat = conn.createStatement();
+        stat.execute("CREATE TABLE TEST(ID BIGINT PRIMARY KEY, T TIMESTAMP WITH TIME ZONE) "
+                + "AS VALUES (1, CURRENT_TIMESTAMP)");
+        PreparedStatement prep = conn.prepareStatement("SELECT T = ANY(SELECT CAST(? AS TIMESTAMP)) FROM TEST");
+        prep.setObject(1, LocalDateTime.now());
+        ResultSet rs = prep.executeQuery();
+        assertTrue(rs.next());
+        assertFalse(rs.getBoolean(1));
+        stat.execute("DROP TABLE TEST");
     }
 
 }

